saml/src/main/java/com/linecorp/armeria/saml/HttpRedirectBindingUtil.java

```diff
@@ -0,0 +1,292 @@
+            throw new SAMLException("no signature parameter in a redirected message");
+        }
+
+        final List<String> sigAlg = parameters.get(PARAM_SIGNATURE_ALGORITHM);
+        if (sigAlg == null || sigAlg.isEmpty()) {
+            throw new SAMLException("no signature algorithm in a redirected message");
+        }
+
+        // The order is one of the followings:
+        // - SAMLRequest={value}&RelayState={value}=SigAlg={value}
+        // - SAMLResponse={value}&RelayState={value}=SigAlg={value}
+        final QueryStringEncoder encoder = new QueryStringEncoder("/");
+        encoder.addParam(messageParamName, parameters.get(messageParamName).get(0));
+
+        final List<String> relayState = parameters.get(PARAM_RELAY_STATE);
+        if (relayState != null && !relayState.isEmpty()) {
+            encoder.addParam(PARAM_RELAY_STATE, relayState.get(0));
+        }
+        encoder.addParam(PARAM_SIGNATURE_ALGORITHM, sigAlg.get(0));
+
+        try {
+            final byte[] input = encoder.toUri().getRawQuery().getBytes(StandardCharsets.UTF_8);
+            if (!XMLSigningUtil.verifyWithURI(validationCredential, sigAlg.get(0),
+                                              Base64Support.decode(signature.get(0)), input)) {
+                throw new SAMLException("signature validation failed");
+            }
+        } catch (SecurityException | URISyntaxException e) {
+            throw new SAMLException("cannot validate signature", e);
+        }
+    }
+
+    /**
+     * Generates a signature of the specified {@code input}.
+     */
+    @VisibleForTesting
+    static String generateSignature(Credential signingCredential,
+                                    String algorithmURI, String input) throws MessageEncodingException {
+        try {
+            final byte[] rawSignature =
+                    XMLSigningUtil.signWithURI(signingCredential, algorithmURI,
+                                               input.getBytes(StandardCharsets.UTF_8));
+            return Base64Support.encode(rawSignature, Base64Support.UNCHUNKED);
+        } catch (SecurityException e) {
+            throw new MessageEncodingException("Unable to sign URL query string", e);
+        }
+    }
+
+    /**
+     * Marshals, deflates and encodes the specified {@code message} to base64-encoded form.
+     */
+    static String deflateAndBase64Encode(SAMLObject message) {
+        try {
+            final String messageStr = nodeToString(XMLObjectSupport.marshall(message));
+            final ByteArrayOutputStream bytesOut = new ByteArrayOutputStream();
+            final Deflater deflater = new Deflater(Deflater.DEFLATED, true);
+            final DeflaterOutputStream deflaterStream = new DeflaterOutputStream(bytesOut, deflater);
+            deflaterStream.write(messageStr.getBytes(StandardCharsets.UTF_8));
+            deflaterStream.finish();
+
+            return Base64Support.encode(bytesOut.toByteArray(), Base64Support.UNCHUNKED);
```

Comment:
As I asked before, I'm not too sure if this `Base64Support` is special, but if you can use JDK, you can use `Base64.Encoder/Decoder.wrap`.